sips/verifiable_consensus_data.md

```diff
@@ -0,0 +1,284 @@
+func verifyAssignedBlockHeader(blockHeader *phase0.BeaconBlockHeader, proposerDuties []ethApi.ProposerDuty) error {
+    for _, proposerDuty := range proposerDuties {
+        if proposerDuty.Slot == blockHeader.Slot {
+            if proposerDuty.ValidatorIndex == blockHeader.ProposerIndex {
+                return nil
+            }
+            else {
+                return errors.New("blockheader proof has unexpected proposer index")
+            }
+        }
+    }
+    return errors.New("blockheader proof's slot is not in last 2 epochs")
+}
+
+//beaconChecks ensures that the beacon chain will accept the attestation
+func beaconChecks(cd *VerifiableConsensusData) error {
+        attestationData, headProof, targetProof, sourceProof, err := cd.GetAttestationData() // error checked in cd.validate()
+        
+        if cd.Duty.Slot != attestationData.Slot {
+            return errors.New("attestation data slot != duty slot")
+        }
+
+        if cd.Duty.CommitteeIndex != attestationData.Index {
+            return errors.New("attestation data CommitteeIndex != duty CommitteeIndex")
+        }
+
+        if attestationData.Target.Epoch < network.EstimatedCurrentEpoch()-1 {
+            return errors.New("attestation data target epoch is into far past")
+        }
+
+        // Addition
+        if attestationData.Target.Epoch != getBeaconNode().EpochFromSlot(headProof.Message.Slot) {
+            return errors.New("Target epoch should be the same as the head vote epoch")
+        }
+        
+        //Addition
+        if attestationData.Target.Epoch != getBeaconNode().EpochFromSlot(targetProof.Message.Slot) {
+            return errors.New("Target epoch doesn't match proof")
+        }
+        
+
+        if attestationData.Source.Epoch >= attestationData.Target.Epoch {
+            return errors.New("attestation data source > target")
+        }
+        
+        // Addition
+        if attestationData.Source.Epoch != getBeaconNode().EpochFromSlot(sourceProof.Message.Slot) {
+            return errors.New("Source epoch doesn't match proof")
+        }
+        
+        // Addition 
+        return isSourceJustified(attestationData.Source))
+}
+
+// isSourceJustified checks against the beacon node if the source is justified. Part of beaconChecks
+func isSourceJustified(attestationData)) error {
+    currentEpoch :=  network.EstimatedCurrentEpoch() 
+    // https://ethereum.github.io/beacon-APIs/#/Beacon/getStateFinalityCheckpoints
+    checkpoints := getBeaconNode().getFinalityCheckpoints()
+    if attestationData.Target.Epoch == currentEpoch {
+        if attestationData.Source.Root != checkpoints.data.currentJustifiedRoot {
+            return Errors.New("source is not expected currently justified root")
+        }
+    } else if attestationData.Source.Root!= checkpoints.data.previousJustifiedRoot {
+        return Errors.New("source is not expected previously justified root")
+    }
+
+    return nil
+}
```

Comment:
This checks exists in [phase0](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#attestations).

In [deneb](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/beacon-chain.md#modified-get_attestation_participation_flag_indices) it is hidden inside the `participation flags` that check eligibility for rewards. There is still an `assert` statement for this check.